### DISCUSSION TITLE
Variance simulations on different computer architectures

### DISCUSSION THREAD

**Discussion by pg320** [nlmixr2/rxode2#561] [Q&A]
I have a run a simple RxODE 2 simulation with population variance on two different computer architectures, but using the same version of R and RStudio and same package version. RxSetSeed has been set to the same value on both computers. 

Should the simulations return the same values? Is there a way of ensuring that they are the same on different computer architectures?

**Comment by mattfidler** [comment]
In short (and in theory) to have the same random deviates, you need to use the same compiler between the platforms.  The easiest is to try to use `gcc` on a mac since windows and linux both use `gcc` by default.  These could also depend on a different compiler version, so I guess you would also have to match the compiler version as well (which means you likely have to match whatever compiler that the windows CRAN uses).

Unfortunately, I haven't tested this, but I believe this to be true, I will explain why you may observe differences.

The simulations have to use a parallel compatible random number generator, since `rxode2` solves each id in parallel:

https://cran.r-project.org/web/packages/sitmo/index.html

The uniform random number generators in Mac, Windows and Linux are the same regardless of if you use `clang`, `gcc` or any other compiler.

The more complex random number generators use the uniform number generator to create a normal random number, or any other random numbers.  These are implemented in C++.  For `rnorm()` for instance the code linked below:

https://github.com/nlmixr2/rxode2random/blob/8ee6b1f64b973d5edb0ccecec696b72a7305366c/src/rxthreefry.cpp#L1270-L1274

These deviates are calculated in the `std::` from the `C++` compiler.  This means the random deviates are controlled by the compiler.

Since rounding, and other functions are treated differently between different compilers, I simply say there are platform differences and really to reproduce the exact same results you need to have the same type of computer too.  

You could use something like docker to reproduce the results.

**Comment by mattfidler** [comment]
You could also use the non-thread safe R simulations which are stable between platforms (but not thread safe).

**Comment by mattfidler** [comment]
As a note, this will be changing in the next release of `rxode2random`.  The random numbers will be based on `boost::random` instead of `std::`.  Since this is an external library, the random deviates will be the same. 

This was pointed out by @rstub  and is a great idea.  Thank you.
